### PR TITLE
Create an ExprHList with operations for zipping multiple expressions into an HList and converting to a product type

### PR DESCRIPTION
**Pull request by jeffmay**
- Add simple Expr.Convert and .as method for converting HList values
- ExprHList for capturing the types of an HList of Exprs
  - Add ExprHList for combining expression operations
  - Add ZipToShortest for type-level computation
  - Use ZipToShortest to implement toHList and zipToHList DSL operations
  - Add support for converting an I ~:> HList into an I ~:> P for any product type that can be converted by shapeless.Generic
  - Add SimpleZipToHListSpec
  - Add SimpleJustifiedZipToHListSpec
  - Add SimpleConcatToHListSpec
  - Add SimpleJustifiedConcatToHListSpec
  - Skip implementation in StandardEngine
  - Add Expr.ProxyVisitor
- Split up DslImplicits into separate traits and files to make it easier to refactor
  - Clean up instances for wrapped versus unwrapped implicits
  - Fix unit tests and add more justified specs
- Add WrapArityMethods for HList to size 9
- Unflatten Justified Traverse definition
- Add more documentation
- Add unit tests to SimpleJustifiedDebuggingSpec

### CODE REVIEW

File: core-v1/src/main/scala/data/Justified.scala [hunk 1]
```diff
@@ -125,14 +126,66 @@ object Justified {
...
+  }
+
+  implicit val semigroupal: Semigroupal[Justified] = new Semigroupal[Justified] {
+    override def product[A, B](
+      fa: Justified[A],
+      fb: Justified[B],
+    ): Justified[(A, B)] = {
+      Justified.byInference("product", (fa.value, fb.value), NonEmptyList.of(fa, fb))
+    }
+  }
```
Comment (jeffmay) [inline core-v1/src/main/scala/data/Justified.scala]: Remove this. It isn't necessary and is probably wrong.

File: core-v1/src/main/scala/dsl/WrapExprDsl.scala [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+//    a: I ~:> W[A],
+//    b: I ~:> W[B],
+//  ): ExprHL2[I, W[A] :: W[B] :: HNil, OP] = a :: b :: ExprHNil2[OP]
+//
+//  def wrap[I, A, B, C](
+//    a: I ~:> W[A],
+//    b: I ~:> W[B],
+//    c: I ~:> W[C],
+//  ): ExprHL2[I, W[A] :: W[B] :: W[C] :: HNil, OP] = a :: b :: c :: ExprHNil2[OP]
+//}
```
Comment (jeffmay) [inline core-v1/src/main/scala/dsl/WrapExprDsl.scala]: I don't think this syntax is necessary since I was able to get `::` working... however, it might be good to add for backwards compatibility. I should either remove this commented code or add backwards compatibility depending on whether the syntax is helpful.

File: core-v1/src/main/scala/engine/StandardEngine.scala [hunk 3]
```diff
@@ -25,6 +25,10 @@ object StandardEngine {
     def apply[PO](state: ExprState[Any, PO]): Visitor[PO, OP] = new Visitor(state)
   }
 
+//  type VisitorParam[PO, OP[_]] = {
+//    type Fn[-I, +O] = PO <:< I => ExprResult[PO, I, O, OP]
+//  }
```
Comment (jeffmay) [inline core-v1/src/main/scala/engine/StandardEngine.scala]: Remove

File: core-v1/src/main/scala/engine/StandardEngine.scala [hunk 4]
```diff
@@ -260,5 +264,16 @@ object StandardEngine {
...
+    override def visitZipToShortestHList[I, F[+_], WL <: HList, UL <: HList](
+      expr: Expr.ZipToShortestHList[I, F, WL, UL, OP],
+    )(implicit
+      zip: ZipToShortest.Aux[F, WL, OP, UL],
+      opO: OP[F[UL]],
+    ): PO <:< I => ExprResult[PO, I, F[UL], OP] = { implicit evPOisI =>
+      // TODO: Figure out how to create an Arrow for this type? Maybe a simpler visitor? Maybe I can rewrite this visitor?
+      val output = zip.zipToShortestWith(expr.exprHList, this)(???)(implicitly)
+      output
+    }
```
Comment (jeffmay) [inline core-v1/src/main/scala/engine/StandardEngine.scala:299]: I am skipping on implementing this because I don't plan on using the standard interpreter for the first milestone release.

File: core-v1/src/main/scala/logic/Conjunction.scala [hunk 5]
```diff
@@ -5,7 +5,7 @@ package logic
 import shapeless.Id
 
 // TODO: Should I allow short-circuiting?
-trait Conjunction[F[_], B, OP[_]] {
+trait Conjunction[F[_], B, -OP[_]] {
```
Comment (jeffmay) [inline core-v1/src/main/scala/logic/Conjunction.scala]: Revert this change.

File: core/src/main/scala/vapors/algebra/HExprList.scala [hunk 6]
```diff
@@ -100,14 +100,14 @@ final case class ExprHCons[V, M[_], H, T <: HList, P](
...
           .alignWith(hs, ts) {
             case Ior.Both(h, t) => Some(h :: t)
             case _ => None
           }
-          .collect {
-            case Some(hlist) => hlist
-          }
+        x.collect {
+          case Some(hlist) => hlist
+        }
```
Comment (jeffmay) [inline core/src/main/scala/vapors/algebra/HExprList.scala]: Revert

File: core-v1/src/main/scala/algebra/Expr.scala [hunk 7]
```diff
@@ -38,6 +41,13 @@ import scala.annotation.nowarn
   */
 sealed abstract class Expr[-I, +O : OP, OP[_]](val name: String) extends Product with Equals {
 
+  /**
+    * TODO: Document better
+    *
+    * @param head the expression to prepend to the constructed [[ExprHList]]
+    */
+  def ::[CI <: I, H](head: Expr[CI, H, OP]): ExprHList[CI, H :: O :: HNil, OP] = head :: this :: ExprHNil[OP]
```
Comment (jeffmay) [inline core-v1/src/main/scala/algebra/Expr.scala]: Do this TODO

File: core-v1/src/main/scala/dsl/IsExprHCons.scala [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+
+      override def head[I, OP[_]](xhl: ExprHList[I, L, OP]): Expr[I, H0, OP] = xhl match {
+        case ExprHCons(head, _) => head.asInstanceOf[Expr[I, H0, OP]]
+        case _ => unexpected
+      }
+
+      override def tail[I, OP[_]](xhl: ExprHList[I, L, OP]): ExprHList[I, T0, OP] = xhl match {
+        case ExprHCons(_, tail) => tail.asInstanceOf[ExprHList[I, T0, OP]]
+        case _ => unexpected
+      }
```
Comment (jeffmay) [inline core-v1/src/main/scala/dsl/IsExprHCons.scala:57]: I should document the danger of this... and why I have determined this to be safe.

File: core-v1/src/main/scala/dsl/StandardRunDsl.scala [hunk 9]
```diff
@@ -37,7 +37,7 @@ trait StandardRunDsl extends RunExprDsl {
     override def run(factTable: FactTable = FactTable.empty): ExprResult[Nothing, Nothing, O, OP] = super.run(factTable)
   }
 
-  final class RunWithStandardExpr[-I, +O](expr: I ~:> O) extends RunWithExpr(expr) {
+  final class RunWithStandardExpr[-I, +O](expr: I ~:> O) extends RunWithExpr[I, O](expr) {
```
Comment (jeffmay) [inline core-v1/src/main/scala/dsl/StandardRunDsl.scala]: Revert

File: core-v1/src/main/scala/dsl/package.scala [hunk 10]
```diff
@@ -0,0 +1,6 @@
+package com.rallyhealth.vapors.v1
+
+package object dsl {
+
+  type CW[C[+_], W[+_], +A] = C[W[A]]
```
Comment (jeffmay) [inline core-v1/src/main/scala/dsl/package.scala:11]: BLKR: Document this type.

File: core-v1/src/main/scala/dsl/ZipToShortest.scala [hunk 11]
```diff
@@ -0,0 +1,168 @@
...
+import algebra.Expr
+
+import cats.arrow.Arrow
+import cats.data.Ior
+import cats.implicits._
+import cats.{Align, Functor, FunctorFilter, Semigroupal}
+import shapeless.{::, HList, HNil, Id}
+
+/**
+  * Interprets an [[ExprHList]] using an [[Arrow]] to produce an F-wrapped [[HList]] of unwrapped results.
```
Comment (jeffmay) [inline core-v1/src/main/scala/dsl/ZipToShortest.scala:10]:
```suggestion
  * Interprets an [[ExprHList]] using an [[Arrow]] to produce a [[W]] wrapped [[HList]] of unwrapped results.
```

File: core-v1/src/test/scala/SimpleDebuggingSpec.scala [hunk 12]
```diff
@@ -240,4 +240,41 @@ class SimpleDebuggingSpec extends FunSuite with CommonDebuggingSpec {
       val o: Option[NestedSelectable] = state.output
     }
   }
+
+  // TODO: These tests would be more useful in a wrapped DSL
+  private val zipToHListExpr0 = ("a".const :: 1.const).toHList
+  private val zipToHListOutput0 = "a" :: 1 :: HNil
```
Comment (jeffmay) [inline core-v1/src/test/scala/SimpleDebuggingSpec.scala]: Copy this test to the wrapped SimpleJustifiedDebuggingSpec

File: core-v1/src/main/scala/algebra/Expr.scala [hunk 13]
```diff
@@ -326,6 +341,111 @@ object Expr {
...
+    def visitZipToShortestHList[I, W[+_], WL <: HList, UL <: HList](
+      expr: Expr.ZipToShortestHList[I, W, WL, UL, OP],
+    )(implicit
+      zip: ZipToShortest.Aux[W, WL, OP, UL],
+      opO: OP[W[UL]],
+    ): I ~:> W[UL]
+  }
+
+  abstract class ProxyVisitor[G[-_, +_], H[-_, +_], OP[_]](protected val underlying: Visitor[G, OP])
+    extends Visitor[H, OP] {
```
Comment (jeffmay) [inline core-v1/src/main/scala/algebra/Expr.scala]: This should be a `class` and not an `abstract class` to avoid issues with missing proxy methods.

**Review by vinit-garg — APPROVED** [review]

**Review by darmbrus — APPROVED** [review]

File: core-v1/src/main/scala/data/Justified.scala [hunk 14]
```diff
@@ -293,6 +293,46 @@ object Justified {
...
+      val derived = f(fa.value)
+      Justified.byInference("map", derived, NonEmptySeq.of(fa))
+    }
+
+    override def traverse[G[_] : Applicative, A, B](fa: Justified[A])(f: A => G[B]): G[Justified[B]] = {
+      val gb = f(fa.value)
+      Applicative[G].map(gb) { b =>
+        Justified.byInference("traverse", b, NonEmptySeq.of(fa))
+      }
+    }
```
Comment (jeffmay) [inline core-v1/src/main/scala/data/Justified.scala:317]: These are very generic names for justification... I wish I could avoid them to be more specific, but the recursive nature of the type definition of `HList` and thus `ExprHList` means that I can only build this chain of justification one node at a time, so we end up with a tree of "map" inferences for each `Expr` node of the `ExprHList`. The best I can do is avoid using this `Functor` / `Traverse` instance for anything other than the building of `HList` outputs and then I can rename `"map"` to something more specific to `HList` like `"::"` or `"cons"` or `"concat"`. For now, I'm gonna keep the generic definition and work backwards to replace anything it is used for later.
